app/assets/javascripts/execution_environments.js

```diff
@@ -1,7 +1,137 @@
+    var version_select = $('.version-select');
+    $(version_select).empty();
+    $(version_select).attr("disabled", false);
+});
+
+
+var setup_programming_language = (function() {
+
+    var proglang_select = $('.proglang-select');
+    var version_select = $('.version-select');
+
+    proglang_select.change(function(){
+        var option = $(proglang_select).children(':selected').text();
+        $.ajax({
+            type: "GET",
+            url: "/programming_languages/versions",
+            data: {proglang: option},
+            dataType: 'json',
+            success: function (data) {
+                // the next thing you want to do
+                update_version_select();
+                for (var i = 0; i < data.length; i++) {
+                    version_select.append('<option value=' + data[i].version + '>' + data[i].version + '</option>');
+                }
+                version_select.val('');
+                version_select.trigger("chosen:updated")
+            }
+        });
+    });
+
+    version_select.change(function(){
+        $('.add-button').attr("disabled", false);
+    });
+});
+
+var setup_new_options = (function($obj) {
+    var select_box = $obj.chosen().data('chosen');
+    select_box.search_field.on('keyup', function (e) {
+        if (event.keyCode === 13) {
+            var option = $("<option>").val(this.value).text(this.value);
+            // Add the new option
+            $($obj).prepend(option);
+            // Automatically select it
+            $($obj).find(option).prop('selected', true).change();
+            // Trigger the update
+            $obj.trigger("chosen:close");
+            $obj.trigger("chosen:updated");
+        }
+    });
+});
+
+var setup_chosen_fields = (function() {
+    var proglang_select = $('.proglang-select');
+    proglang_select.chosen({
+        disable_search_threshold: 0,
+        no_results_text: "No results match. Press enter to add"
+    });
+    var version_select = $('.version-select');
+    version_select.chosen({
+        disable_search_threshold: -1,
+        no_results_text: "No results match. Press enter to add"
+    });
+
+    setup_new_options(proglang_select);
+    setup_new_options(version_select);
+});
+
+var show_error_message = (function(error) {
+    $('<div class="proglang-error alert alert-danger fade in">'+ error +'</div>')
+        .insertAfter('.prog-lang-form')
+        .delay(3000)
+        .slideUp('medium')
+        .queue(function() {
+            $(this).remove();
+        });
+});
+
+var show_default = (function($show, $hide){
+    $hide.hide();
+    $show.show();
+});
+
 $(function() {
   if ($.isController('execution_environments')) {
     if ($('.edit_execution_environment, .new_execution_environment').isPresent()) {
       new MarkdownEditor('#execution_environment_help');
+      setup_programming_language();
+      //show_programming_language_values();
+      setup_chosen_fields();
     }
   }
 });
+
```

Comment:
this is called also outside of the view / controller of execution_environments.... I suppose this is not intended?

If it is - why?

Reply:
I don't really understand what you mean.
setup_programming_language() and setup_chosen_fields() is only supposed to be called when loading the form.
But I thought that's the point of these lines: 

if ($.isController('execution_environments')) {
    if ($('.edit_execution_environment, .new_execution_environment').isPresent())

Am I mistaken? If so, how do I make sure it is only called with the form?

Reply:
I think the comment refers to the block below, lines 95-137. 😄

Reply:
Yeah, I saw it now, has been fixed :D